Add hardlink support for Linux & macOS

BlauFx opened the pull request:
https://user-images.githubusercontent.com/34138240/209813281-0548784f-35db-4349-be90-56caf5f1b3b1.mp4

BlauFx committed:
Add hardlink support for Linux

BlauFx committed:
Remove blank line

bdach commented on osu.Game/IO/HardLinkHelper.cs:
```diff
@@ -29,9 +28,16 @@ public static bool CheckAvailability(string testDestinationPath, string testSour
+
+                bool isHardLinkAvailable = false;
+
+                if (RuntimeInfo.OS == RuntimeInfo.Platform.Windows)
+                    isHardLinkAvailable = CreateHardLink(testDestinationPath, testSourcePath, IntPtr.Zero);
+                else if (RuntimeInfo.OS == RuntimeInfo.Platform.Linux)
+                    isHardLinkAvailable = link(testSourcePath, testDestinationPath) == 0;
```
Can this become its own method now that it has multiple platform implementations? Kinda feels bad seeing this same platform choice code in 2 places (here and `RealmFileStore`).

bdach commented on osu.Game/IO/HardLinkHelper.cs:
```diff
@@ -70,6 +76,9 @@ public static int GetFileLinkCount(string filePath)
             return result;
         }
 
+        [DllImport("libc",  SetLastError = true)]
```
```suggestion
        [DllImport("libc", SetLastError = true)]
```

bdach commented:
Also this message on the first run setup overlay is a bit wrong now:

https://github.com/ppy/osu/blob/e4c060fc3afa1991fc1e05b79f9134c10713f306/osu.Game/Overlays/FirstRunSetup/ScreenImportFromStable.cs#L139-L140

BlauFx committed:
Fix formatiing

BlauFx committed:
Mention the filesystem should be NTFS on Windows

BlauFx committed:
Move duplicated code into its own method

BlauFx committed:
Remove unnecessary directive

BlauFx committed:
Change name of variable

bdach committed:
Revert inlining of hard link creation into condition

Just feels bad. Mixing data access with actual underlying logic.

bdach committed:
Use switch statement in `AttemptHardLink()`

bdach committed:
Rename and xmldoc hard link creation method

bdach committed:
Change warning message about file duplication

- It was being glued in an ugly way that would have prevented sanely
  localising it.

- Even on Linux, the filesystem (whichever one the user has chosen out
  of the multitude available) still needs to support hard links for them
  to have a chance of working.

bdach committed:
Implement `GetFileLinkCount()` for Linux

bdach committed:
Fix broken condition

bdach committed:
Fix broken spacing

bdach commented:
I started going through this pull again but ended up pretty much rewriting it for multiple reasons elaborated in commit messages. I don't think this was actually tested either because there was an egregious logic bug that meant that this not only didn't work on linux, but also broke windows (49b0ec9ddb43d0f507ba8155d7bec69ed7876bb8).

Going to need a second review from someone. And going to need to re-test hard links on windows too to make sure nothing is broken.

(As an aside I do wonder if macOS can just use the same linux implementation...)

bdach committed:
Disable naming rule inspection on struct stat definition

peppy committed:
Merge branch 'master' into add_linux_hardlink

peppy committed:
Also support hard links on macOS

peppy commented:
This looks to work fine on macOS so I've added support.

bdach commented on osu.Game/Overlays/FirstRunSetup/ScreenImportFromStable.cs:
```diff
@@ -132,12 +132,13 @@ private void updateStablePath()
                 copyInformation.AddLink("Learn more about how \"hard links\" work", LinkAction.OpenWiki, @"Client/Release_stream/Lazer/File_storage#via-hard-links");
             }
-            else if (RuntimeInfo.OS != RuntimeInfo.Platform.Windows)
+            else if (RuntimeInfo.OS != RuntimeInfo.Platform.Windows && RuntimeInfo.OS != RuntimeInfo.Platform.Linux)
```
@peppy I think this is going to need updating to `IsDesktop` too? Would entail trying to migrate across filesystems.

peppy committed:
Fix incorrect hard link validity check in stable import screen

bdach reviewed (approved):
Re-tested windows. Looks fine to me.
